File: DFS/图论/863. 二叉树中所有距离为 K 的结点.py
""" 
先构造一个连通图，每一个节点，作为key，value是节点的相邻节点列表
图建好之后，用 BFS 去遍历
过程中需要一个 visited 集合去存已经遍历过的节点，避免重复算

作者：niconiconi-12
链接：https://leetcode.cn/problems/all-nodes-distance-k-in-binary-tree/solution/chi-xiao-dou-python-bfs-hashmap-dai-ma-j-d3kl/
来源：力扣（LeetCode）
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。
"""



class Solution(object):
    def distanceK(self, root, target, k):
        self.graph = {}
        
        def buildGraph(currentNode):
            if currentNode.left == None and currentNode.right == None:
                if currentNode.val not in self.graph:
                    self.graph[currentNode.val] = []

            
            if currentNode.left:
                if currentNode.val not in self.graph:
                    self.graph[currentNode.val] = []
                self.graph[currentNode.val].append(currentNode.left.val)
                self.graph[currentNode.left.val] = []
                self.graph[currentNode.left.val].append(currentNode.val)
                buildGraph(currentNode.left)
            
            if currentNode.right:
                if currentNode.val not in self.graph:
                    self.graph[currentNode.val] = []
                self.graph[currentNode.val].append(currentNode.right.val)
                self.graph[currentNode.right.val] = []
                self.graph[currentNode.right.val].append(currentNode.val)
                buildGraph(currentNode.right)
        buildGraph(root)

        def findNodes(target, k):
            from collections import deque
            queue = deque([target.val])
            visited = set()
            counter = 0
            result = []

            while queue:
                queue_len = len(queue)
                
                if counter == k:
                    for idx in range(queue_len):
                        currentVal = queue.popleft()
                        if currentVal not in visited:
                            result.append(currentVal)
                    return result

                for _ in range(queue_len):
                    currentVal = queue.popleft()
                    if currentVal not in visited:
                        visited.add(currentVal)
                        if self.graph[currentVal]:
                            queue.extend(self.graph[currentVal][:])

                counter += 1

            return result

        return findNodes(target, k)

class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
root = TreeNode(3)
lnode1 = TreeNode(5)
lnode2 = TreeNode(6)
lnode3 = TreeNode(2)
lnode4 = TreeNode(7)
lnode5 = TreeNode(4)

rnode1 = TreeNode(1)
rnode2 = TreeNode(0)
rnode3 = TreeNode(8)

root.left = lnode1
lnode1.left = lnode2
lnode1.right= lnode3
lnode3.left = lnode4
lnode3.right = lnode5

root.right = rnode1
rnode1.left = rnode2
rnode1.right = rnode3
# root = TreeNode(1)
s =  Solution()
res = s.distanceK(root, lnode1, 2)
print(res)

File: DFS/Others/863. 二叉树中所有距离为 K 的结点.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * 题意：返回与 target 节点距离为 k 的所有节点值
 * 思路：
 * 1. 将二叉树转换成无向图（每条边都是双向的）
 * 2. 从 target 出发执行 BFS，找到距离为 k 的所有节点
 */

var distanceK = function (root, target, k) {
    const graph = new Map(); // Map<number, number[]>，键和值都用节点的 val 表示

    /**
     * 使用 BFS 建立无向图
     * 每个节点与其左右子节点相连（双向边）
     */
    function buildGraph(node) {
        if (!node) return;
        const queue = [node];

        while (queue.length) {
            const cur = queue.shift(); // 当前节点
            const v = cur.val; // 当前节点值

            // 确保当前节点在图中有一个空邻接表
            if (!graph.has(v)) graph.set(v, []);

            // 处理左子节点
            if (cur.left) {
                queue.push(cur.left);
                const lv = cur.left.val; // 左子节点值
                if (!graph.has(lv)) graph.set(lv, []);
                // 双向连接
                graph.get(v).push(lv);
                graph.get(lv).push(v);
            }

            // 处理右子节点
            if (cur.right) {
                queue.push(cur.right);
                const rv = cur.right.val; // 右子节点值
                if (!graph.has(rv)) graph.set(rv, []);
                // 双向连接
                graph.get(v).push(rv);
                graph.get(rv).push(v);
            }
        }
    }

    // 第一步：建图
    buildGraph(root);

    // 第二步：从 target 节点出发执行 BFS
    const t = target.val; // 注意：用节点值而不是节点对象
    if (!graph.has(t)) return [];

    // 特殊情况：k = 0，直接返回自身
    if (k === 0) return [t];

    const visited = new Set([t]); // 记录访问过的节点值
    let queue = [t]; // 当前层队列（节点值）

    // BFS，层序遍历直到距离为 k
    while (k > 0 && queue.length) {
        const size = queue.length;
        k--;
        for (let i = 0; i < size; i++) {
            const cur = queue.shift();
            for (const nxt of graph.get(cur)) {
                // 如果这个邻居节点没访问过，加入队列
                if (!visited.has(nxt)) {
                    visited.add(nxt);
                    queue.push(nxt);
                }
            }
        }
        // 当 k 变为 0 时，当前队列中的节点即为目标节点
        if (k === 0) return queue;
    }

    // 如果没有符合条件的节点，返回空数组
    return [];
};
